x-pack/plugin/rollup/src/main/java/org/elasticsearch/xpack/rollup/mapper/AggregateMetricFieldMapper.java

```diff
@@ -0,0 +1,340 @@
+                return new Explicit<>(metrics, true);
+            }
+            return Defaults.METRICS;
+        }
+
+        @Override
+        public AggregateMetricFieldMapper build(BuilderContext context) {
+            setupFieldType(context);
+            return new AggregateMetricFieldMapper(name, fieldType, defaultFieldType,
+                context.indexSettings(), multiFieldsBuilder.build(this, context),
+                ignoreMalformed(context), metrics(context), copyTo);
+        }
+    }
+
+    public static class TypeParser implements Mapper.TypeParser {
+
+        @Override
+        public Mapper.Builder<Builder, AggregateMetricFieldMapper> parse(String name,
+                                                                         Map<String, Object> node,
+                                                                         ParserContext parserContext) throws MapperParsingException {
+            AggregateMetricFieldMapper.Builder builder = new AggregateMetricFieldMapper.Builder(name);
+            TypeParsers.parseField(builder, name, node, parserContext);
+            for (Iterator<Map.Entry<String, Object>> iterator = node.entrySet().iterator(); iterator.hasNext(); ) {
+                Map.Entry<String, Object> entry = iterator.next();
+                String propName = entry.getKey();
+                Object propNode = entry.getValue();
+                if (propName.equals(Names.METRICS.getPreferredName())) {
+                    String metricsStr[] = XContentMapValues.nodeStringArrayValue(propNode);
+                    // Convert the array of Metric enum to a Set of their string representation
+                    Set<String> supporteddMetrics = Arrays.stream(Metrics.values()).map(s -> s.toString()).collect(Collectors.toSet());
+
+                    // Make sure that metrics are supported
+                    Set<String> parsedMetrics = new LinkedHashSet<>();
+                    for (int i = 0; i < metricsStr.length; i++) {
+                        if (supporteddMetrics.contains(metricsStr[i]) == false) {
+                            throw new MapperParsingException("Metric [" + metricsStr[i] + "] is not supported.");
+                        } else {
+                            parsedMetrics.add(metricsStr[i]);
+                        }
+                    }
+                    builder.metrics(parsedMetrics);
+                    iterator.remove();
+                } else if (propName.equals(Names.IGNORE_MALFORMED)) {
+                    builder.ignoreMalformed(XContentMapValues.nodeBooleanValue(propNode, name + "." + Names.IGNORE_MALFORMED));
+                    iterator.remove();
+                } else if (TypeParsers.parseMultiField(builder, name, parserContext, propName, propNode)) {
+                    iterator.remove();
+                }
+            }
+            return builder;
+        }
+    }
+
+    public static final class AggregateMetricFieldType extends MappedFieldType {
```

Comment:
I think you can use a plain `NumericFieldType` instance for each metric and delegate the execution based on the context ? For now this would all delegate to the default metric and we think at how metrics aggs should communicate a different choice as a follow up.